### PR TITLE
Various smaller fixes in the Frontend & Pylint config

### PR DESCRIPTION
**Pull request by dime10**
`.pylintrc` has been reworked to remove any unnecessary configuration and only include exceptions relevant to this project.
The following checks are disabled globally:
- `too-few-public-methods`: Catalyst code frequently makes use of classes holding data or representing an object without any methods
- `invalid-name`: this check disallows short variable names like 'x'
- `too-many-locals`: current coding style employs local variables liberally
- `cyclic-import`: currently the Catalyst package contains harmless cyclical imports
- `import-error`: since the CI doesn't necessarily have access to all modules imported by Catalyst, and legitimate import errors would show up during tests anyways

All issues that appeared due to the reconfiguration in the main `catalyst` package have been fixed, existing issues in auxiliary packages like `test` will be ignored.

Adds `isort` for automatic sorting of python imports to our formatting checks and make command.

Also fixes smaller issues in the frontend including:
 - emitting errors around compilation pass outputs & aborting in the absence of llvmir
 - enable `get_output_of` with custom pass pipelines
 - gitignore newly generated shared lib
 - cleanup files in `keep_intermediate` test

### CODE REVIEW

**Comment by erick-xanadu** [comment]
@dime10 maybe a suggestion, we currently emit an error if the extension of files to be passed to quantum opt don't match. I say that this is a weak guarantee and also hinders the ability to customize the pipeline. Maybe just catch the error when the subprocess fails? Just because an extension matches doesn't mean that the contents of the file is what is expected. Likewise, just because an extension doesn't match it doesn't mean that the contents of the file are wrong. What do you think?

**Comment by dime10** [comment]
> @dime10 maybe a suggestion, we currently emit an error if the extension of files to be passed to quantum opt don't match. I say that this is a weak guarantee and also hinders the ability to customize the pipeline. Maybe just catch the error when the subprocess fails? Just because an extension matches doesn't mean that the contents of the file is what is expected. Likewise, just because an extension doesn't match it doesn't mean that the contents of the file are wrong. What do you think?

That's probably a good idea now that don't have a single pipeline anymore.

File: frontend/catalyst/compiler.py [hunk 1]
```diff
@@ -395,19 +391,19 @@ def run(infile, outfile=None, flags=None, fallback_compilers=None, options=None)
...
+        msg += "Please provide a compatible compiler via $CATALYST_CC."
         raise EnvironmentError(msg)
 
 
 class Compiler:
     """Compiles MLIR modules to shared objects."""
 
     def __init__(self):
-        self.pass_pipeline_output = None
+        self.pass_pipeline_output = {}
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compiler.py:403]: Thanks!

[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/compiler.py:403]: Well I figured on a subsequent run we probably wouldn't want the old entries around, or is that irrelevant?

[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/compiler.py:403]: You are right. I expected the Compiler object to be regenerated but it appears that is not the case.

File: frontend/test/pytest/test_compiler.py [hunk 2]
```diff
@@ -322,6 +322,36 @@ def get_output_filename(infile):
...
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def workflow():
+            qml.X(wires=1)
+            return qml.state()
+
+        mlir_module, _, _ = get_mlir(workflow)
+        compiler = Compiler()
+        compiler.run(mlir_module, CompileOptions(pipelines=[MyPass]))
+        result = compiler.get_output_of("MyPass")
+        assert result == "hi\n"
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_compiler.py:355]: Nice test. This is a TODO for me later but I would like to also encapsulate the python -> JAXPR into a pipeline. That way, one wouldn't need to depend on the mlir module in this test.

**Review by erick-xanadu — APPROVED** [review]
Looks awesome. If you are done with it I approve it. I know there are complains about the coverage, but it looks good to me. Do you know why the coverage is complaining?

**Comment by dime10** [comment]
> Looks awesome. If you are done with it I approve it. I know there are complains about the coverage, but it looks good to me. Do you know why the coverage is complaining?

Thank you!
Given that the patch is at 100% percent I think this could be a case of deleted lines leading to a lower coverage percentage overall 🤦

**Pull request merged**